bluesky_widgets/jupyter/_search_input.py

```diff
@@ -0,0 +1,180 @@
+                self.until_widget.value = now + self.model.until
+
+    def _on_since_view_changed(self, change):
+        self.model.since = change["new"]
+
+    def _on_since_model_changed(self, event):
+        now = self._now()
+        if isinstance(event.date, timedelta):
+            new_datetime = now + event.date
+            if event.date == timedelta(hours=-1):
+                self.radio_button_group.index = self.radio_button_group.options.index("1 Hour")
+            elif event.date == timedelta(days=-1):
+                self.radio_button_group.index = self.radio_button_group.options.index("24 Hours")
+            elif event.date == timedelta(days=-7):
+                self.radio_button_group.index = self.radio_button_group.options.index("1 Week")
+            elif event.date == timedelta(days=-30):
+                self.radio_button_group.index = self.radio_button_group.options.index("30 Days")
+            elif event.date == timedelta(days=-365):
+                self.radio_button_group.index = self.radio_button_group.options.index("1 Year")
+            else:
+                pass
+        else:
+            if event.date == GRACE_HOPPER_BIRTHDAY:
+                self.radio_button_group.index = self.radio_button_group.options.index("All")
+            else:
+                self.radio_button_group.index = None
+            new_datetime = event.date
+        with _blocked(self.since_widget, self._on_since_view_changed, names="value"):
+            self.since_widget.value = new_datetime
+        with _blocked(self.until_widget, self._on_until_view_changed, names="value"):
+            self.until_widget.value = now
+
+    def _on_until_view_changed(self, change):
+        self.model.until = change["new"]
+
+    def _on_until_model_changed(self, event):
+        if not isinstance(event.date, timedelta):
+            self.radio_button_group.index = None
+            with _blocked(self.until_widget, self._on_until_view_changed, names="value"):
+                self.until_widget.value = event.date
+        else:
+            with _blocked(self.until_widget, self._on_until_view_changed, names="value"):
+                self.until_widget.value = event.date + self._now()
+
+    def _on_text_view_changed(self, change):
+        self.model.text = change["new"]
+
+    def _on_text_model_changed(self, event):
+        self.text_search_input.value = event.text
+
+    def _on_field_search_updated(self, event):
+        for k, v in event.update.items():
+            text_widget = self._fields[k]
+            text_widget.value = v
+
+    @staticmethod
+    def _now():
+        return datetime.now(tz=LOCAL_TIMEZONE).replace(second=0, microsecond=0)
+
+
+@contextlib.contextmanager
+def _blocked(widget, handler, names=All, type="change"):
+    "Block signals from this object inside the context."
+    widget.unobserve(handler, names, type)
+    yield
+    widget.observe(handler, names, type)
```

Comment:
We should ensure that the `handler` is restored (i.e.`observe(..)` is run) even if an error is raised by the code run during the `yield` stage. See https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager for example.